### DISCUSSION TITLE
Stuck trying to implement htmx todo list example with Jstachio fragments

### DISCUSSION THREAD

**Discussion by Vislike** [jstachio/jstachio#415] [Q&A] [answered]
Hi,

I am trying to convert this example: https://github.com/AussieGuy0/java-htmx-todo/ but instead of j2html using Jstachio.
It uses a `record Todo(String id, String content, boolean completed)`

What i have so far is this mustache file:
```
{{<htmlBase}}
{{$body}}
	<h1>Todo App</h1>
	{{$todolist}}
	<ul id="todo-list">
		{{#todos}}
		<li>
			{{content}} {{id}}
		</li>
		{{/todos}}
	</ul>
	{{/todolist}}
{{/body}}
{{/htmlBase}}
```

And these Jstache records:
```java
	@JStache(path = "taskList")
	public record TaskListPage(List<Todo> todos) {
	}
 
	@JStache(path = "taskList#todolist")
	public record TodoList(List<Todo> todos) {
	}
```

This works so far, i can hit my main endpoint and get the entire html page, i can also hit the todos endpoint and get only the "ul" list

But the problem starts in the next step by adding {{$todoItem}}.
Mustache:
```
{{<htmlBase}}
{{$body}}
	<h1>Todo App</h1>
	{{$todolist}}
	<ul id="todo-list">
		{{#todos}}
		{{$todoItem}}
		<li>
			{{content}} {{id}}
		</li>
		{{/todoItem}}
		{{/todos}}
	</ul>
	{{/todolist}}
{{/body}}
{{/htmlBase}}
```

Java:
```java
	@JStache(path = "taskList")
	public record TaskListPage(List<Todo> todos) {
	}
 
	@JStache(path = "taskList#todolist")
	public record TodoList(List<Todo> todos) {
	}
 
	@JStache(path = "taskList#todoItem")
	public record TodoItem(Todo todo) {
	}
```

Now i can't compile taskList#todoItem since it can't find the variabel 'content' and this can be fixed in the mustache file by writing {{todo.content}}, but then taskList and taskList#todolist don't compile since they don't understand todo variabel.

What would you consider the best practice to solve this?

I have tested to change the TodoItem to the parts i need from Todo. But that results in a lot of parameter copying at the endpoint.

Another way that works is to take the Todo record and annotate that with @JStache. But is that a good idea? I don't know if i should mix the "Database model" and the "View model".

I am thinking if there are some way to not explode out the parts of Todo when iterating over the list instead getting the result as a composite object. But at the same time i don't know how to get a "handle" to it then, something like {{#todos as todo}}. I have had a similar problem when experimenting with lambdas e.g. if i return a List<Todo> from the lambda if i access it as {{.}} it is a list, so i do {{#.}}{{/.}} then again i don't get the todo object, it instead explodes out to its parts. But maybe that is the most logical way to do it, and i have just gotten the concepts mixed up in my head.

I would love to hear the opinion from someone with more experience with html templates.

Edit:
In my test project i went with a separate view record i choose this path since i also needed a view only variabel "editing" that should not be part of the "database". But i have no idea if this is considered good practice or not, but it at least works.

The project can be found here: https://github.com/Vislike/java-htmx-todo-jstachio

**Comment by agentgt** [comment]
Sorry for the delayed response! I am going to look into this later today and should have an update for you this weekend.

**Comment by agentgt** [comment, marked as answer]
The easiest solution is to add methods to `TodoItem`.

The key thing is not to be afraid to decorate the annotated models (and no I don't mean the actual domain object `Todo`)

```java
	@JStache(path = "taskList#todoItem")
	public record TodoItem(Todo todo) {
		public String content() { return todo.content(); }
		public UUID id() { return todo.id(); }
	}
```

I know that above example does seem boilerplate and not DRY but you are trying to mesh two different shape things.

One way to approach JStachio is a more top down approach where you write the templates and use variable names that make sense for the context of the template. Then in the annotated models you adapt to those variable names instead of trying to make the template adapt to your domain model.

**Reply by Vislike** [reply]
Great suggestions, sounds like a good approach to handle the situation. I would love to get a more top down mindset, and i think this could help. Next time i will try to design the mustache file first.
